kv: include locking strength and durability in Get/Scan/RevScan SafeFormat

lyang24 opened the pull request:
kv: include locking strength and durability in Get/Scan/RevScan SafeFormat

The goal of this pr is improving observability. We are adding lock strength
and lock durability with Get/Scan/RevScan request if the request is locking.
This is implemented by introducing an optional extension interface to Request
Interface called SafeFormatterRequest. We are also refactoring inside
BatchRequest.SafeFormat with the added interface. Please note the subtle
changed introduced here: if the EndKey is not present we print only Key with
square brackets, and this applies to all the request types.

Fixes: https://github.com/cockroachdb/cockroach/issues/114475
Release note: None.

Commit:
kv: include locking strength and durability in Get/Scan/RevScan SafeFormat

The goal of this pr is improving observability. We are adding lock strength
and lock durability with Get/Scan/RevScan request if the request is locking.
This is implemented by introducing an optional extension interface to Request
Interface called SafeFormatterRequest. We are also refactoring inside
BatchRequest.SafeFormat with the added interface. Please note the subtle
changed introduced here: if the EndKey is not present we print only Key with
square brackets, and this applies to all the request types.

Fixes: #114475

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114481)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 1 of 2 files at r1, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Njn69pm6XGvkCw8IJrl):*
Same two comments about the commit message as in https://github.com/cockroachdb/cockroach/pull/114636.
___
*[`pkg/kv/kvpb/batch.go` line 813 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Njn6IEe40SQtJiUd-E4:-Njn6IEf3h85H-tmDZoM:b-d1pd3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/01618a8d11e63ec9856624cc7dc1a712fef03b89/pkg/kv/kvpb/batch.go#L813)):*
> ```Go
> 
> 		req := arg.GetInner()
> 		if et, ok := req.(*EndTxnRequest); ok {
> ```

I was thinking that this new `SafeFormatterRequest` interface could be used for all of these special cases. So instead, we'd just have some logic that looked like
```go
if safeFormatterReq, ok := req.(SafeFormatterRequest); ok {
    safeFormatterReq.SafeFormat(s, verb)
} else {
    s.Print(req.Method())
}
if len(h.EndKey) > 0 {
    s.Printf(" [%s,%s)", h.Key, h.EndKey)
} else {
    s.Printf(" [%s]", h.Key)
}
```


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
Thanks for the feedback, i have addressed the comments. Please note the subtle change that we are applying this logic at the end of request types now.
```
if len(h.EndKey) > 0 {
	s.Printf(" [%s,%s)", h.Key, h.EndKey)
} else {
	s.Printf(" [%s]", h.Key)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-:-NkEFZNl4X6E8_PVDCe0:b71wrjw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvpb/batch.go` line 813 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Njn6IEe40SQtJiUd-E4:-NkEFIvtEIPGWGyjB10p:b8r0j1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/01618a8d11e63ec9856624cc7dc1a712fef03b89/pkg/kv/kvpb/batch.go#L813)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I was thinking that this new `SafeFormatterRequest` interface could be used for all of these special cases. So instead, we'd just have some logic that looked like
```go
if safeFormatterReq, ok := req.(SafeFormatterRequest); ok {
    safeFormatterReq.SafeFormat(s, verb)
} else {
    s.Print(req.Method())
}
if len(h.EndKey) > 0 {
    s.Printf(" [%s,%s)", h.Key, h.EndKey)
} else {
    s.Printf(" [%s]", h.Key)
}
```
</blockquote></details>

Thanks it is so much cleaner after the refactor.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/kv/kvpb/api.go` line 197 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-NkbHJ5wEUpIlGGFljP6:-NkbHJ5wEUpIlGGFljP7:b-e3q52y) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e256ce99f1084192f1b05144ee3f9089f2add0/pkg/kv/kvpb/api.go#L197)):*
> ```Go
> }
> 
> // SafeFormatterRequest is an optional extension interface used to allow request to do custom formatting.
> ```

It would be nice to add some request-specific tests for these requests which are implementing the `SafeFormatterRequest` interface. We could add those small unit tests in `pkg/kv/kvpb/string_test.go`, right below `TestBatchRequestString`.

When doing so, we can test the safe-string and redacted-string forms. `TestReplicaStringAndSafeFormat` is an example of how to do that.
___
*[`pkg/kv/kvpb/api.go` line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-NkbIOBZ9niPkec2cLFq:-NkbIOBZ9niPkec2cLFr:b33aale) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e256ce99f1084192f1b05144ee3f9089f2add0/pkg/kv/kvpb/api.go#L206)):*
> ```Go
> 
> // SafeFormat implements the redact.SafeFormatter interface.
> func (gr GetRequest) SafeFormat(s redact.SafePrinter, _ rune) {
> ```

These methods should all use pointer receivers.
___
*[`pkg/kv/kvpb/api.go` line 211 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-NkbIXVE91psC8XhhapK:-NkbIXVE91psC8XhhapL:bhwsxfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e256ce99f1084192f1b05144ee3f9089f2add0/pkg/kv/kvpb/api.go#L211)):*
> ```Go
> 		return
> 	}
> 	s.Printf("[lockStrength=%s,lockDurability=%s]", gr.KeyLockingStrength, gr.KeyLockingDurability)
> ```

Here and below: we can make this less verbose. How about something like:
```go
s.Printf("(%s,%s)", gr.KeyLockingStrength, gr.KeyLockingDurability)
```
___
*[`pkg/kv/kvpb/batch.go` line 813 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Njn6IEe40SQtJiUd-E4:-NkbJiTR-05ul3wAIi13:b8rlt5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/01618a8d11e63ec9856624cc7dc1a712fef03b89/pkg/kv/kvpb/batch.go#L813)):*
<details><summary><i>Previously, lyang24 (Lanqing Yang) wrote…</i></summary><blockquote>

Thanks it is so much cleaner after the refactor.
</blockquote></details>

Agreed, this did turn out nicely!


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvpb/api.go` line 197 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-NkbHJ5wEUpIlGGFljP6:-NkbpN3e0vOc6g9NpzLj:b-628kcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e256ce99f1084192f1b05144ee3f9089f2add0/pkg/kv/kvpb/api.go#L197)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It would be nice to add some request-specific tests for these requests which are implementing the `SafeFormatterRequest` interface. We could add those small unit tests in `pkg/kv/kvpb/string_test.go`, right below `TestBatchRequestString`.

When doing so, we can test the safe-string and redacted-string forms. `TestReplicaStringAndSafeFormat` is an example of how to do that.
</blockquote></details>

yep added the `TestRequestSafeFormat` the keys are printed in the batch request SafeFormat method so it wont display here just fyi. 
___
*[`pkg/kv/kvpb/api.go` line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-NkbIOBZ9niPkec2cLFq:-NkbpFVPBSV2Av7hxkef:b-j92saq) ([raw file](https://github.com/cockroachdb/cockroach/blob/20e256ce99f1084192f1b05144ee3f9089f2add0/pkg/kv/kvpb/api.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These methods should all use pointer receivers.
</blockquote></details>

great catch!


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lyang24)

___
*[`pkg/kv/kvpb/string_test.go` line 103 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Nkufr0n-1OO7l1aWaUz:-Nkufr0n-1OO7l1aWaV-:b-n2h5ws) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c57c68eee6d294e7384cf3d5de9bbf13ee8dad/pkg/kv/kvpb/string_test.go#L103)):*
> ```Go
> // Unit test the requests that implemented SafeFormatterRequest interface.
> func TestRequestSafeFormat(t *testing.T) {
> 	gr := &kvpb.GetRequest{
> ```

Can we turn this into a table-driven test to make it easier to add new request types? Something like:
```go
testCases := []struct {
	req        Request
	redactable string
	redacted   string
}{
	{
		req: &kvpb.GetRequest{
			RequestHeader: kvpb.RequestHeader{
				Key: roachpb.Key("a"),
			},
			KeyLockingStrength:   lock.Shared,
			KeyLockingDurability: lock.Unreplicated,
		},
		redactable: "Get(Shared,Unreplicated)",	
		redacted: "Get(Shared,Unreplicated)",	
	},
	// ...
}
for _, c := range testCases {
	t.Run(c.Method(), func(t *testing.T) {
		// ...
	})
}
```
___
*[`pkg/kv/kvpb/string_test.go` line 170 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Nkuf-oKBSV0UY5dypl0:-Nkuf-oKBSV0UY5dypl1:b699eqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c57c68eee6d294e7384cf3d5de9bbf13ee8dad/pkg/kv/kvpb/string_test.go#L170)):*
> ```Go
> 		PusheeTxn: pusheeTxn.TxnMeta,
> 	}
> 	require.EqualValues(t, "PushTxn(‹PUSH_TIMESTAMP›,00fbff58->00fbff59)", redact.Sprint(ptr))
> ```

The redaction of `PUSH_TIMESTAMP` is interesting here.

We should add a new commit which makes `PushTxnType` implement `redact.SafeValue`.

To do that, we'll want a line like the following to `api.go`:
```go
// SafeValue implements the redact.SafeValue interface.
func (PushTxnType) SafeValue() {}
```
And then we'll have to add an entry to `pkg/testutils/lint/passes/redactcheck/redactcheck.go`.


<!-- Sent from Reviewable.io -->

lyang24 committed:
kv: implement redact.SafeValue for PushTxnType

This commit implements redact.SafeValue for PushTxnType with the goal of
displaying the PushTxnType on PushTxnRequest with the redacted format.

Fixes: #114475
Release note: None.

lyang24 reviewed (commented):
Thank you for reviewing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-:-NkwIyTD11iIZCLuxlGi:bzh25ef)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvpb/string_test.go` line 103 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Nkufr0n-1OO7l1aWaUz:-NkwIi0o9rgVEL3QP1X8:bs822mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c57c68eee6d294e7384cf3d5de9bbf13ee8dad/pkg/kv/kvpb/string_test.go#L103)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we turn this into a table-driven test to make it easier to add new request types? Something like:
```go
testCases := []struct {
	req        Request
	redactable string
	redacted   string
}{
	{
		req: &kvpb.GetRequest{
			RequestHeader: kvpb.RequestHeader{
				Key: roachpb.Key("a"),
			},
			KeyLockingStrength:   lock.Shared,
			KeyLockingDurability: lock.Unreplicated,
		},
		redactable: "Get(Shared,Unreplicated)",	
		redacted: "Get(Shared,Unreplicated)",	
	},
	// ...
}
for _, c := range testCases {
	t.Run(c.Method(), func(t *testing.T) {
		// ...
	})
}
```
</blockquote></details>

Thank you this is a nice pattern looking forward to incorporate in my future developments.
___
*[`pkg/kv/kvpb/string_test.go` line 170 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-Nkuf-oKBSV0UY5dypl0:-NkwIs5652lK0zfdo6OS:b-q0fx7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6c57c68eee6d294e7384cf3d5de9bbf13ee8dad/pkg/kv/kvpb/string_test.go#L170)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The redaction of `PUSH_TIMESTAMP` is interesting here.

We should add a new commit which makes `PushTxnType` implement `redact.SafeValue`.

To do that, we'll want a line like the following to `api.go`:
```go
// SafeValue implements the redact.SafeValue interface.
func (PushTxnType) SafeValue() {}
```
And then we'll have to add an entry to `pkg/testutils/lint/passes/redactcheck/redactcheck.go`.
</blockquote></details>

Thank you for catching this. I fixed it.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

bors r=nvanbenschoten

> Reviewed 1 of 1 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114481#-:-Nl5-stL3Z9mMpdx9Ap3:b-qdhl0k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lyang24)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r-

nvanbenschoten commented:
@lyang24 this PR skewed with https://github.com/cockroachdb/cockroach/pull/115282/commits/e53cc10829fa7740de7b40bfed5e5496e8bc4e26. Could you rebase it on master?

nvanbenschoten commented:
bors r=nvanbenschoten

nvanbenschoten commented:
bors r=nvanbenschoten

michae2 commented:
This is really helpful! Thank you!

michae2 commented:
@lyang24 @nvanbenschoten I'm going to backport this to 23.2. I think it would be really helpful to have in v23.2.1+.

blathers backport release-23.2
